firedrake/interpolation.py

```diff
@@ -381,6 +392,27 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
         return copyin + (parloop_compute_callable, ) + copyout
 
 
+try:
+    _expr_cachedir = os.environ["FIREDRAKE_TSFC_KERNEL_CACHE_DIR"]
+except KeyError:
+    _expr_cachedir = os.path.join(tempfile.gettempdir(),
+                                  f"firedrake-tsfc-expression-kernel-cache-uid{os.getuid()}")
+
+
+def _compile_expression_key(comm, expr, to_element, ufl_element, domain, parameters):
+    """Generate a cache key suitable for :func:`tsfc.compile_expression_dual_evaluation`."""
+    # Since the caching is collective, this function must return a 2-tuple of
+    # the form (comm, key) where comm is the communicator the cache is collective over.
+    # FIXME FInAT elements are not safely hashable so we ignore them here
+    key = hash(expr), hash(ufl_element), hash(domain), utils.tuplify(parameters)
```

Comment:
Is this the right key? For a disk key, I think you want a hash that is stable over coefficient and index renumbering (which would come from ufl.algorithms.compute_expression_signature or similar, rather than hash).

I am also not sure that the domain needs to be hashed. Remember, this needs to cover the information that is necessary for the kernel, not the global iteration?

Reply:
@wence- I've had a look at this and I'm not completely sure how to accomplish what you want. Would something like `slow_key` found in [assemble_expressions.py](https://github.com/firedrakeproject/firedrake/blob/master/firedrake/assemble_expressions.py#L310) work?

Reply:
I think UFL has a `compute_expression_signature` somewhere. The main thing is to provide renumberings for coefficients and domains (`compute_form_signature` just gets the coefficients and domain numberings and then passes down to `compute_expression_signature` IIRC).

Reply:
This is now done. Sorry for taking so long over this.